int64 total_length
int64 current_node